### ISSUE TITLE
Schedule: loadEvents offset error when specifying clientTimezone

### ISSUE THREAD

**Issue by dwagmuse** [primefaces/primefaces#2977]
When using the clientTimezone attribute in schedule component to force the view to display times in a fixed timezone the fetchEvents callback to the server that supports lazy loading is passing incorrect arguments. The selection window seems to be offset by the value of the browser default (local) timezone.  It appears to work correctly when clientTimezone is set to the default 'local'.
In this case, I'm trying to force the browser to display in UTC.
See proposed fix at the end.

## 1) Environment
- PrimeFaces version: 6.1 (ALL, through 6.1.8)
- Does it work on the newest released PrimeFaces version? Version? No
- Does it work on the newest sources in GitHub? No
- Application server + version: Tomcat 8/8.5
- Affected browsers: all

## 2) Expected behavior
Build schedule view to use lazy loading.
Using clientTimezone to force view to display events in UTC. Go to today, and select the day mode.
The callback to loadEvents passes (start,end) parameters which I expect should bound a window of 00:00:00 to 24:00:00 for the selected day. 
...

## 3) Actual behavior
The actual values are offset by the value of browser local TZ.  
..

## 4) Steps to reproduce
Using the example bean and view, just open the view on the default day (today), select the day view button on top right to issue an events query for today's events.
.. 

## 5) Sample XHTML
    <!DOCTYPE html>
    <html xmlns="http://www.w3.org/1999/xhtml"
	xmlns:h="http://java.sun.com/jsf/html"
	xmlns:f="http://java.sun.com/jsf/core"
	xmlns:p="http://primefaces.org/ui"
	xmlns:ui="http://java.sun.com/jsf/facelets">
    <h:head>
	<title>Schedule Demo</title>
    </h:head>
    <h:body>
    <h:form id="stest">
    <p:growl id="messages" showDetail="true" />
    <h:panelGrid columnClasses="value">
   		
        <h3>Editable Lazy Schedule </h3>
        <h:outputText value="#{scheduleBean.tzone}"/>
        <p:schedule id="schedule2"
        	value="#{scheduleBean.lazyEventModel}" 
        	widgetVar="myschedule2" 
        	timeZone="#{scheduleBean.tzone}"
        	clientTimeZone="#{scheduleBean.tzone}"
        	scrollTime="#{scheduleBean.scrolltime}"
			minTime="#{scheduleBean.mintime}" 
			maxTime="#{scheduleBean.maxtime}">
			
			<p:ajax event="dateSelect" 
				listener="#{scheduleBean.onDateSelect}" 
				update="eventDetails" 
				oncomplete="PF('eventDialog').show();" />
            <p:ajax event="eventSelect" 
            	listener="#{scheduleBean.onEventSelect}" 
            	update="eventDetails" 
            	oncomplete="PF('eventDialog').show();" />
            <p:ajax event="eventMove" 
            	listener="#{scheduleBean.onEventMove}" 
            	update="messages" />
            <p:ajax event="eventResize" 
            	listener="#{scheduleBean.onEventResize}" 
            	update="messages" />
 
		</p:schedule>
		
    </h:panelGrid>
 
    <p:dialog widgetVar="eventDialog" header="Event Details" showEffect="clip" hideEffect="clip">
        <h:panelGrid id="eventDetails" columns="2">
            <p:outputLabel for="title" value="Titles:" />
            <p:inputText id="title" value="#{scheduleBean.event.title}" 
            	required="true" />
 
            <p:outputLabel for="from" value="From:" />
            <p:calendar id="from" value="#{scheduleBean.event.startDate}" 
            	timeZone="#{scheduleBean.tzone}" pattern="dd/MM/yyyy HH:mm"/>
 
            <p:outputLabel for="to" value="To:" />
            <p:calendar id="to" value="#{scheduleBean.event.endDate}" 
            	timeZone="#{scheduleBean.tzone}" pattern="dd/MM/yyyy HH:mm"/>
 
            <p:outputLabel for="allDay" value="All Day:" />
            <h:selectBooleanCheckbox id="allDay" 
            	value="#{scheduleBean.event.allDay}" />
 
            <p:commandButton type="reset" value="Reset" />
            <p:commandButton id="addButton" value="Save" 
            	actionListener="#{scheduleBean.addEvent}" oncomplete="PF('myschedule2').update();PF('eventDialog').hide();" />
        </h:panelGrid>
    </p:dialog>
    
    </h:form>
 
    <style type="text/css">
        .value {
            width: 1000px;
        }
    </style>

    </h:body>
    </html>
..

## 6) Sample bean
(modified version of showcase demo)

    package schedbug;

    import java.io.Serializable;
    import java.text.SimpleDateFormat;
    import java.util.Calendar;
    import java.util.Date;
    import java.util.TimeZone;

    import javax.annotation.PostConstruct;
    import javax.faces.application.FacesMessage;
    import javax.faces.bean.ManagedBean;
    import javax.faces.bean.ViewScoped;
    import javax.faces.context.FacesContext;
    import javax.faces.event.ActionEvent;

    import org.primefaces.event.ScheduleEntryMoveEvent;
    import org.primefaces.event.ScheduleEntryResizeEvent;
    import org.primefaces.event.SelectEvent;
    import org.primefaces.model.DefaultScheduleEvent;
    import org.primefaces.model.DefaultScheduleModel;
    import org.primefaces.model.LazyScheduleModel;
    import org.primefaces.model.ScheduleEvent;
    import org.primefaces.model.ScheduleModel;

    @ManagedBean
    @ViewScoped
    public class ScheduleBean implements Serializable {
	private static final long serialVersionUID = 1L;
	private ScheduleModel lazyEventModel;
	
	private TimeZone localtz= TimeZone.getDefault();
	private TimeZone utc=TimeZone.getTimeZone("UTC");
	// The timezone parameter is only used in
	// the server side to convert events to JSON string. Not sure
	// why anybody would want to change that, because it seems to have
	// no effect on the browser behavior. The js side parses the string
	// and adjusts to client tz regardless of what zone the JSON string
	// is rendered in.
	private String tzname=utc.getID();
	// Client timezone is the zone in which events are displayed in the
	// browser. The default is 'local' which would use the default local
	// zone on the browser host. That's likely to be what most users want
	// most of the time. 
	private String clienttzname=tzname;
	
	// not clear what this property does. Documentation is
	// ambiguous, but if this is used a deprecation warning says to use scrolltime.
	private String firsthour="0"; // default is 6
	// these properties appear to only affect the display. changing the
	// seems to have no effect on loadEvents parameters.
	private String mintime="00:00:00"; 
	private String maxtime="23:00:00";
	private String scrolltime="12:00:00";

	private ScheduleEvent event = new DefaultScheduleEvent();

	@PostConstruct
	public void init() {
		System.out.println("ScheduleBean.init() tz="+tzname);
		System.out.println("local default tz="+localtz.getID());
		
		lazyEventModel = new LazyScheduleModel() {
			private static final long serialVersionUID = 1L;

			private String dump(Date t) {
				return format(t)+" "+t.getTime();
			}
			private void debug(Date start, Date end) {
				System.out.println("***\n");
				System.out.println("loadEvents start="+dump(start));
				System.out.println("loadEvents end  ="+dump(end));
				
				Date now = new Date();
				System.out.println("loadEvents now  ="+dump(now));
				
				Calendar c = Calendar.getInstance(utc);
				c.setTime(now);
				// compute what start/end of day should look like
				c.set(Calendar.HOUR_OF_DAY, 0);
				c.set(Calendar.MINUTE, 0);
				c.set(Calendar.SECOND, 0);
				c.set(Calendar.MILLISECOND, 0);
				Date expected=c.getTime();
				System.out.println("start today="+dump(expected));
				c.add(Calendar.HOUR, 24);
				System.out.println("end today  ="+dump(c.getTime()));
				long error=expected.getTime()-start.getTime();
				System.out.println("error="+error+" ms ="+error/3600000+"h");
				
			}
			
			@Override
			public void loadEvents(Date start, Date end) {
				debug(start, end);
				
				
				// ScheduleRenderer translates the _start and _end params coming from
				// client directly into Date objects. So if there is any error in
				// the values, it came from the js client.
				
				Date random = getRandomDate(start);
				addEvent(new DefaultScheduleEvent("Lazy Event 1", random, random));

				random = getRandomDate(start);
				addEvent(new DefaultScheduleEvent("Lazy Event 2", random, random));
				
				// These events are all explicitly created in UTC so that
				// I can be sure about the times. However, I'm ignoring
				// the query window, so some of these will be out of range.
				
				addEvent(new DefaultScheduleEvent("Yesterday 8pm ", previousDay8Pm(), previousDay11Pm()));
				addEvent(new DefaultScheduleEvent("Today 1pm ", today1Pm(), today6Pm()));
				addEvent(new DefaultScheduleEvent("Tomorrow 9am-11am", nextDay9Am(), nextDay11Am()));
				addEvent(new DefaultScheduleEvent("Plant the new garden stuff", theDayAfter3Pm(), fourDaysLater3pm()));

			}
		};
	}

	public Date getRandomDate(Date base) {
		Calendar date = Calendar.getInstance(utc);
		date.setTime(base);
		date.add(Calendar.DATE, ((int) (Math.random() * 30)) + 1); // set random
																	// day of
																	// month

		return date.getTime();
	}

	public Date getInitialDate() {
		Calendar calendar = Calendar.getInstance(utc);
		calendar.set(calendar.get(Calendar.YEAR), Calendar.FEBRUARY, calendar.get(Calendar.DATE), 0, 0, 0);

		return calendar.getTime();
	}

	public ScheduleModel getLazyEventModel() {
		return lazyEventModel;
	}

	private Calendar today() {
		Calendar calendar = Calendar.getInstance(utc);
		calendar.set(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DATE), 0, 0, 0);

		return calendar;
	}

	private Date previousDay8Pm() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.AM_PM, Calendar.PM);
		t.set(Calendar.DATE, t.get(Calendar.DATE) - 1);
		t.set(Calendar.HOUR, 8);

		return t.getTime();
	}

	private Date previousDay11Pm() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.AM_PM, Calendar.PM);
		t.set(Calendar.DATE, t.get(Calendar.DATE) - 1);
		t.set(Calendar.HOUR, 11);

		return t.getTime();
	}

	private Date today1Pm() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.AM_PM, Calendar.PM);
		t.set(Calendar.HOUR, 1);

		return t.getTime();
	}

	private Date theDayAfter3Pm() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.DATE, t.get(Calendar.DATE) + 2);
		t.set(Calendar.AM_PM, Calendar.PM);
		t.set(Calendar.HOUR, 3);

		return t.getTime();
	}

	private Date today6Pm() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.AM_PM, Calendar.PM);
		t.set(Calendar.HOUR, 6);

		return t.getTime();
	}

	private Date nextDay9Am() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.AM_PM, Calendar.AM);
		t.set(Calendar.DATE, t.get(Calendar.DATE) + 1);
		t.set(Calendar.HOUR, 9);

		return t.getTime();
	}

	private Date nextDay11Am() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.AM_PM, Calendar.AM);
		t.set(Calendar.DATE, t.get(Calendar.DATE) + 1);
		t.set(Calendar.HOUR, 11);

		return t.getTime();
	}

	private Date fourDaysLater3pm() {
		Calendar t = (Calendar) today().clone();
		t.set(Calendar.AM_PM, Calendar.PM);
		t.set(Calendar.DATE, t.get(Calendar.DATE) + 4);
		t.set(Calendar.HOUR, 3);

		return t.getTime();
	}

	public ScheduleEvent getEvent() {
		return event;
	}

	public void setEvent(ScheduleEvent event) {
		this.event = event;
	}

	public void onEventSelect(SelectEvent selectEvent) {
		event = (ScheduleEvent) selectEvent.getObject();
		System.out.println("onEventSelect e="+event.getDescription());
	}

	public void onDateSelect(SelectEvent selectEvent) {
		event = new DefaultScheduleEvent("", (Date) selectEvent.getObject(), (Date) selectEvent.getObject());
		System.out.println("onDateSelect start="+format(event.getStartDate()));
	}

	public void onEventMove(ScheduleEntryMoveEvent event) {
		FacesMessage message = new FacesMessage(FacesMessage.SEVERITY_INFO, "Event moved",
				"Day delta:" + event.getDayDelta() + ", Minute delta:" + event.getMinuteDelta());

		addMessage(message);
	}

	public void onEventResize(ScheduleEntryResizeEvent event) {
		FacesMessage message = new FacesMessage(FacesMessage.SEVERITY_INFO, "Event resized",
				"Day delta:" + event.getDayDelta() + ", Minute delta:" + event.getMinuteDelta());

		addMessage(message);
	}

	private void addMessage(FacesMessage message) {
		FacesContext.getCurrentInstance().addMessage(null, message);
	}

	public String getTzone() {
		return tzname;
	}
	public String getClientTzone() {
		return clienttzname;
	}
	
	public String getMintime() {
		return mintime;
	}
	public String getMaxtime() {
		return maxtime;
	}
	public String getScrolltime() {
		return scrolltime;
	}
	public String getFirsthour() {
		return firsthour;
	}
	
	private String format(Date d) {
		SimpleDateFormat df1 = new SimpleDateFormat("YYYY-MM-dd HH:mm:ss Z");
		df1.setTimeZone(utc);
		return df1.format(d);
	}
    }

..

I traced the code into schedule.js setupEventSource method. The problem appears to be that it is using the default zone offset to convert start/end values before issuing the fetch query.  The following patch seems to fix the problem:

    setupEventSource: function() {
        var $this = this;
        //offset = moment().zone()*60000; // BUG: not correct if clientTimezone is set to something other than local

        this.cfg.events = function(start, end, timezone, callback) {
            var offset = start.zone()*60000;// <-- FIX: assume start,end in same zone
            var options = {
                source: $this.id,
                process: $this.id,
                update: $this.id,
                formId: $this.cfg.formId,
                params: [
                    {name: $this.id + '_start', value: start.valueOf() + offset},
                    {name: $this.id + '_end', value: end.valueOf() + offset}
                ],
...

**Comment by melloware** [comment]
When I looked at the code in the trunk it looks different than yours.  Its always using UTC for the offset.

```javascript
    setupEventSource: function() {
        var $this = this,
        offset = moment().utcOffset()*60000; //<-- Different than yours above
```

So is it possible this is already fixed in master?

It also looked like the moment.js library was updated as part of this change.

**Comment by dwagmuse** [comment]
If you look in the sources to moment.js you will see that the utcOffse()t method is just a replacement for the deprecated zone() method. So that change is a no-op. The problem is that it is taking the offset from the call to moment() which constructs a new moment in default (local) TZ. In my fix I obtain the offset from the start time, which has been set to the configured zone. There's probably a way to get the zone configuration setting directly, but my fix works, and the one in master does not.

**Comment by melloware** [comment]
Got it.  Just wanted to make sure that wasn't fix.  I will fix it.

**Comment by dwagmuse** [comment]
Is there anything I can do to help get this fix merged and released?

**Comment by melloware** [comment]
You will most likely have to wait until the 6.2 release for this fix or build it locally yourself for now.  I doubt it will get merged into 6.1.x Elite release unless you are a PRO subscriber.
